cdap-ui/app/directives/dag-plus/my-dag-ctrl.js

```diff
@@ -275,67 +274,91 @@ angular.module(PKG.name + '.commons')
+          initNormalNode(node);
         }
 
-        vm.instance.makeTarget(node.name, {
-          isTarget: true,
-          dropOptions: { hoverClass: 'drag-hover' },
-          anchor: 'ContinuousLeft',
-          allowLoopback: false
-        });
+        if (!vm.instance.isTarget(node.name)) {
+          let targetOptions = {
+            isTarget: true,
+            dropOptions: { hoverClass: 'drag-hover' },
+            anchor: 'ContinuousLeft',
+            allowLoopback: false
+          };
+          if (node.type === 'alertpublisher') {
+            targetOptions.scope = 'alertScope';
+          } else if (node.type === 'errortransform') {
+            targetOptions.scope = 'errorScope';
+          }
+          vm.instance.makeTarget(node.name, targetOptions);
+        }
       });
     }
 
-    function initNormalNode(nodeName) {
-      let sourceObj = {
-        isSource: true,
-        filter: function(event) {
-          // we need this variable because when the user clicks on the endpoint circle, multiple
-          // 'mousedown' events are fired
-          if (event.target.className.indexOf('endpoint-circle') !== -1 && !endpointClicked) {
-            endpointClicked = true;
-          }
+    function addEndpointForNormalNode(endpointId, customConfig, nodesWithEndpoints, nodeName) {
+      let endpointDOMEl = document.getElementById(endpointId);
+      let endpoint = vm.instance.addEndpoint(endpointDOMEl, Object.assign({}, { isSource: true }, customConfig));
+      addListenersForEndpoint(endpoint, endpointDOMEl);
+      if (nodesWithEndpoints && nodeName) {
+        nodesWithEndpoints.push(nodeName);
+      }
+    }
 
-          return event.target.className.indexOf('endpoint-circle') !== -1;
-        },
-        connectionType: 'basic'
+    function initNormalNode(node) {
+      let nodeName = node.name;
+      const addErrorAlertEndpoints = (node) => {
+        if (vm.shouldShowAlertsPort(node)) {
+          addEndpointForNormalNode(node.name + '_alert', {
+            scope: 'alertScope'
+          });
+        }
+        if (vm.shouldShowErrorsPort(node)) {
+          addEndpointForNormalNode(node.name + '_error', {
+            scope: 'errorScope'
+          });
+        }
       };
-      if (vm.isDisabled) {
-        sourceObj.enabled = false;
-      }
 
-      vm.instance.makeSource(nodeName, sourceObj);
+      if (normalNodes.indexOf(nodeName) === -1) {
+        addEndpointForNormalNode(nodeName + '_endpoint', null, normalNodes, nodeName);
+        if (node.type === 'alertpublisher') {
+          alertNodes.push(nodeName);
+        } else if (node.type === 'errortransform') {
+          errorNodes.push(nodeName);
+        }
+
+        if (_.isEmpty(vm.pluginsMap)) {
+          let subAvailablePluginsForErrorAlerts = AvailablePluginsStore.subscribe(() => {
+            vm.pluginsMap = AvailablePluginsStore.getState().plugins.pluginsMap;
+            if (!_.isEmpty(vm.pluginsMap)) {
```

Comment:
Checking for `vm.pluginsMap` seems to be redundant. Can't we just subscribe to the store and just unsubscribe once we have something in the `vm.pluginsMap` like we do in the inner condition block?